internal/cache/file/cache_handle.go

```diff
@@ -0,0 +1,232 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/file/downloader"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+)
+
+const (
+	InvalidFileHandle      = "destroy this cache_handle: invalid file handle"
+	InvalidFileDownloadJob = "destroy this cache_handle: invalid download job"
+	InvalidCacheHandle     = "destroy this cache_handle: invalid cache handle"
+	InvalidFileInfoCache   = "file info cache is nil"
+	ErrInSeekingFileHandle = "destroy this cache_handle: error while seeking file handle"
+	ErrInReadingFileHandle = "destroy this cache_handle: error while reading file handle"
+	FallbackToGCS          = "read via gcs without destroying this cache_handle"
+	MB                     = 1 << 20
+	MaxAllowedMB           = 8 * MB
+)
+
+type CacheHandle struct {
+	// fileHandle to a local file which contains locally downloaded data.
+	fileHandle *os.File
+
+	//	fileDownloadJob is a reference to async download Job.
+	fileDownloadJob *downloader.Job
+
+	// fileInfoCache contains the reference of fileInfo cache.
+	fileInfoCache *lru.Cache
+
+	// isSequential saves if the current read performed via cache_handle is sequential or
+	// random. If true then sequential otherwise random.
+	isSequential bool
+
+	// prevOffset stores the offset of previous cache_handle read call. This is used
+	// to decide the type of read.
+	prevOffset int64
+
+	// contributesToJobRefCount indicates that the cache handle contains the reference
+	// count of the Job. We maintain this reference count manually, since we keep the
+	// Job reference always, regardless of whether the read is sequential or random.
+	// However, by design, we want to cancel the Job if there is no sequential read
+	// active that depends on the Job.
+	contributesToJobRefCount bool
+}
+
+// NewCacheHandle constructs and returns CacheHandle object.
+func NewCacheHandle(localFileHandle *os.File, fileDownloadJob *downloader.Job, fileInfoCache *lru.Cache, initialOffset int64) *CacheHandle {
+	return &CacheHandle{
+		fileHandle:               localFileHandle,
+		fileDownloadJob:          fileDownloadJob,
+		fileInfoCache:            fileInfoCache,
+		isSequential:             initialOffset == 0,
+		prevOffset:               initialOffset,
+		contributesToJobRefCount: true,
```

Comment:
Just writing down the cases that we need to handle here:
(We need to check file cache in all cases irrespective of type of read)
1. When --download-for-random-read=true:

    -  If read is sequential - wait for download & ignore the job reference count because we don't need to cancel the async job in this case once started.
    - if read is random - don't wait for download & ignore the job reference count because we don't need to cancel the async job in this case once started.
    - if read changes from sequential to random - no handling, follow the random read case.

1. When --download-for-random-read=false:

    - If read is sequential - wait for download & increase the job reference count.
    - if read is random - don't **download** (rather it should check the status of job here - so job should have a function GetJobStatus) & ignore the job reference count because we don't need to cancel the async job in this case.
    - if read changes from sequential to random - don't **download** & decrease the job reference count. if count becomes 0, on file cache handler side, cancel the job. _Note:_ _There is one corner case here - What if count becomes 0 at certain point and then again increases. In that case, file cachehandler should "Reset" the download job i.e. set to NOT_STARTED state._  (I will handle add Reset function in job if all this makes sense to you).
    
If you think handling all the cases above will make things very complicated then - My suggestion is to keep it simple like this:
1. If download-for-random-read = true:
    - if sequential read - wait for download.
    - if random read - don't wait for download.
    - if sequential read changes to random - follow random read.
 
 2. if download-for-random-read = false
    - if sequential read - wait for download.
    - if random read - don't **download** (still requires GetStatus)
    - if sequential read changes to random - follow random read.

Reply:
Yeah, that's what we are doing here. Handling code will be random_reader.go.

It's simple only, just we need to ignore JobCancelling logic via jobRefCount when download-for-random=true.

`Note: There is one corner case here - What if count becomes 0 at certain point and then again increases. In that case, file cachehandler should "Reset" the download job i.e. set to NOT_STARTED state` - yeah sure.

Reply:
As discussed offline, we are going with the simple logic